### ISSUE TITLE
Ansible stalls during setup when pushing to large amount of hosts

### ISSUE THREAD

**Issue by digi691** [ansible/ansible#17152]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

- Performance Improvement (executor, strategy plugin, etc... )
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel c5b5a20031) last updated 2016/08/18 18:11:18 (GMT -500)
  lib/ansible/modules/core: (detached HEAD 91a839f1e3) last updated 2016/08/18 19:24:24 (GMT -500)
  lib/ansible/modules/extras: (detached HEAD 1aeb9f8a8c) last updated 2016/08/18 19:24:24 (GMT -500)

```
##### CONFIGURATION

strategy=free & strategy=linear
##### OS / ENVIRONMENT

Control: Ubuntu 14.04.3 LTS
Managing: N/A
##### SUMMARY

When pushing Ansible to over 1000 hosts, Ansible will stall when running the built-in setup tasks on the managing hosts.  A linear strategy seems to get further than free, but both end up stalling before getting to any actual task in the playbook.
##### STEPS TO REPRODUCE

When checking out commit d2b3b2c03e2934b126f5701e5f6e25821e2dbe35 , Ansible will stall when pushing to over 1000 hosts, when checked out to the prior commit 80385a47bda4207a39cc6437458bbe3c3bb34d63 the behavior does not exist.

```
<playbook_test.yml>

---

- hosts: "{{ hosts|default('all') }}"
  strategy: "{{ strategy|default('linear') }}"
  user: root
  vars_files:
    - var_files/{{ ansible_distribution }}.yml
    - var_files/{{ ansible_distribution }}-{{ ansible_distribution_major_version }}.yml
    - var_files/{{ ansible_distribution }}-{{ ansible_distribution_version }}.yml
  handlers:
    - include: handlers/main.yml
      static: yes
  tasks:
    - name: test
      command: "echo foo"
```

hosts file: https://gist.github.com/digi691/77929ae3c1fd3eec3a23fbda30439862
##### EXPECTED RESULTS

Successful run in the 5-6 minute timeframe.
##### ACTUAL RESULTS

Ansible will stall with strategy free after about 20 TASK [setup] messages have printed to the screen.  strategy linear maybe gets 2-3x more than that before stalling.  Output below is the last few dozen of lines while using -vvvv.

```
<xxxx-xxxxx> PUT /tmp/root/tmpSvqING TO /root/.ansible/tmp/ansible-tmp-1471484820.62-266171058798943/setup.py
<xxxx-xxxxx> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAu
thentication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r '[chhq-vudtbrix09]'
<xxxx-xxxxx> ESTABLISH SSH CONNECTION FOR USER: root
<xxxx-xxxxx> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthenti
cation=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r chhq-vudtbrix09 '/bin/sh -c '"'"'chmod u+x /root/.ansible/tmp/ansible-tmp-1471484820.62-266171058798943/ /root/.ansible/tmp/ansible-tmp-1471484820.62-266171
058798943/setup.py && sleep 0'"'"''
<xxxx-xxxxx> ESTABLISH SSH CONNECTION FOR USER: root
<xxxx-xxxxx> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthenti
cation=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt chhq-vudtbrix09 '/bin/sh -c '"'"'/usr/bin/python /root/.ansible/tmp/ansible-tmp-1471484820.62-266171058798943/setup.py; rm -rf "/root/.ansible/tmp/ansib
le-tmp-1471484820.62-266171058798943/" > /dev/null 2>&1 && sleep 0'"'"''
<xxxx-xxxxx> ESTABLISH SSH CONNECTION FOR USER: root
<xxxx-xxxxx> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentica
tion=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r lnhq-vupadm01 '/bin/sh -c '"'"'chmod u+x /root/.ansible/tmp/ansible-tmp-1471484830.66-44898110337002/ /root/.ansible/tmp/ansible-tmp-1471484830.66-44898110337
002/setup.py && sleep 0'"'"''
<xxxx-xxxxx> ESTABLISH SSH CONNECTION FOR USER: root
<xxxx-xxxxx> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentica
tion=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt lnhq-vupadm01 '/bin/sh -c '"'"'/usr/bin/python /root/.ansible/tmp/ansible-tmp-1471484830.66-44898110337002/setup.py; rm -rf "/root/.ansible/tmp/ansible-tm
p-1471484830.66-44898110337002/" > /dev/null 2>&1 && sleep 0'"'"''
<xxxx-xxxxx> PUT /tmp/root/tmp7z18Xh TO /root/.ansible/tmp/ansible-tmp-1471484830.89-182835333315991/setup.py
<xxxx-xxxxx> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuth
entication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r '[sg1e-vupadm01]'
<xxxx-xxxxx> ESTABLISH SSH CONNECTION FOR USER: root
<xxxx-xxxxx> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentica
tion=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r sg1e-vupadm01 '/bin/sh -c '"'"'chmod u+x /root/.ansible/tmp/ansible-tmp-1471484830.89-182835333315991/ /root/.ansible/tmp/ansible-tmp-1471484830.89-1828353333
15991/setup.py && sleep 0'"'"''
<xxxx-xxxxxx> ESTABLISH SSH CONNECTION FOR USER: root
<xxxx-xxxxxx> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentica
tion=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt sg1e-vupadm01 '/bin/sh -c '"'"'/usr/bin/python /root/.ansible/tmp/ansible-tmp-1471484830.89-182835333315991/setup.py; rm -rf "/root/.ansible/tmp/ansible-t
mp-1471484830.89-182835333315991/" > /dev/null 2>&1 && sleep 0'"'"''
<xxxx-xxxxx> PUT /tmp/root/tmpQ6zoSd TO /root/.ansible/tmp/ansible-tmp-1471484824.79-234080379367754/setup.py
<xxxx-xxxxx> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAu
thentication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r '[mt1n-vudredis02]'
<mt1n-vudredis02> ESTABLISH SSH CONNECTION FOR USER: root
<xxx-xxxxx> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthenti
cation=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r mt1n-vudredis02 '/bin/sh -c '"'"'chmod u+x /root/.ansible/tmp/ansible-tmp-1471484824.79-234080379367754/ /root/.ansible/tmp/ansible-tmp-1471484824.79-234080
379367754/setup.py && sleep 0'"'"''
<mt1n-vudredis02> ESTABLISH SSH CONNECTION FOR USER: root
<xxx-xxxxx2> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthenti
cation=no -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt mt1n-vudredis02 '/bin/sh -c '"'"'/usr/bin/python /root/.ansible/tmp/ansible-tmp-1471484824.79-234080379367754/setup.py; rm -rf "/root/.ansible/tmp/ansib
le-tmp-1471484824.79-234080379367754/" > /dev/null 2>&1 && sleep 0'"'"''

```

**Comment by dagwieers** [comment]
Are you sure this is not related to very slow systems ? I've used Ansible on more than 3000 systems without any issue, but we frequently had systems with issues that could cause Ansible to wait seemingly indefinitely.

Ansible does not report which systems it is connecting to (unless you look at the process list, e.g. pstree), but I like shmux more in this regard, as you can query it while it is executing stuff in parallel. So you can query it on-screen to ask which have not been processed yet, or you can stop a specific connection so it would not hang forever. Maybe that's a simple way for you to detect which system is the culprit and debug the issue ?

**Comment by digi691** [comment]
@jimi-c ,

As you suggested in IRC, I tried playing with the amount of forks.  I started to have mixed results around 85 forks in linear strategy.  At 85 forks in linear strategy I was about 25% successful on a whole push.  The closer I got to 100 forks the more successful I became.  As far as I can tell, at 100 forks I was up to 100% success rate.  This wasn't automated so not totally conclusive.  During the failures the stall was again in setup.  When I switched to the free strategy I could never get past the stall described in the bug report (Tried up to 200 forks).

**Comment by jimi-c** [comment]
@digi691 the `meta_threading` branch was merged into devel today, which includes a change which I believe addresses your issue. Please let me know when you've had a chance to test the most recent version of devel to see if you are still running into this problem.

Thanks!

**Comment by digi691** [comment]
My testing was done as of commit 06c76034dc203dac52ddf5cb6b6afc4b484a25af with 1384 hosts in my host list.

Strategy linear mode is now successful, tested with multiple pushes.
`ansible-playbook -f 25 -i hosts -e strategy=linear playbook-test_setup.yml`

Strategy Free is still unsuccessful although I get a lot further into the play before it stalls, I believe past all of setup, but cannot be sure.  When trying to create the debug output for this comment, the debug stalls a lot sooner than without ANSIBLE_DEBUG=1.  Below are the output from two separate runs, one for when the fork was set to 25 and another where the fork was set to 50.  Although it may not be useful, it seems something else isn't quite right or an indication of another problem.

`ANSIBLE_DEBUG=1 ansible-playbook -f 25 -i hosts -e strategy=free playbook-test_setup.yml | tee -a debug_stall_9-4-16`
https://gist.github.com/digi691/319dff4bf7bb5e6ad8dc4eb8366e1c24

`ANSIBLE_DEBUG=1 ansible-playbook -f 50 -i hosts -e strategy=free playbook-test_setup.yml | tee -a debug_stall_9-4-16_try2`
https://gist.github.com/digi691/e4a751b8ac08a37ceaaa3405f50a6c5e

@jimi-c ,
I'll pop on IRC on Tuesday morning to see if your around to see if you want further details.

**Comment by jimi-c** [comment]
@digi691 another user who's been experiencing a similar problem pointed out it might be a problem with doing a fork from a thread, in which case specifically the debug output lock may be inherited by the forked process in a locked state and then hangs.

I have rebased this branch against devel: https://gist.github.com/jimi-c/69118da22ee9341fddddfb3fe841d84d, which may help. If that doesn't, you may want to try commenting out the acquire/release calls there in debug to see if that resolves the issue for you (in which case we will know definitively that it's the debug lock).

There is a chance I may take Tuesday off, so if I don't respond leave me an update here on this issue.

Thanks!

**Comment by digi691** [comment]
@jimi-c,

I couldn't apply that patch from the gist you posted.  Looks like your role_include.py may still be ahead of devel?  Output is below:

```
 ❯ git apply debug_deadlock_fix.patch --check --recount -v                                                                                                                [11:56:55]
Checking patch lib/ansible/playbook/role_include.py...
error: while searching for:
        self._from_files = {}
        self._parent_role = role


    def get_block_list(self, play=None, variable_manager=None, loader=None):


error: patch failed: lib/ansible/playbook/role_include.py:58
error: lib/ansible/playbook/role_include.py: patch does not apply
```

Since that failed, I then commented out the debug acquire/release calls for debug in utils/display.py as you suggested.  This has fixed the deadlock issue for the debug output.  I then ran the following command:

`ANSIBLE_DEBUG=1 ansible-playbook -f 100 -i hosts -e strategy=free playbook-test_setup.yml | tee -a debug_stall_9-5-16`

This job never seemed to stop, so I prematurely stopped it after the runtime I was seeing on a linear run (~6 minutes) that was forked to 100.  The raw debug file is 324M; 16M in a tarball.  How do you want me to get this debug output file to you?

**Comment by jimi-c** [comment]
@digi691 sorry for the long delay in following up, but the background worker to read results has been merged in for a while, and I haven't heard any further complaints regarding stalls. If you have any further questions or concerns let me know, otherwise I'll close this issue in a few days.

Thanks!

**Comment by digi691** [comment]
@jimi-c I've been busy as well...  Though good timing as I was running into issues pushing our ssh keys role to our whole environment last night to all our hosts on Ansible 2.1.1.  So I checked out Ansible 2.2.0 to see if the stalling issue was fixed, unfortunately it still stalls out.  I switched to the simple test playbook I posted earlier in this bug report and that stalled as well under Ansible 2.2.0.  I'm still trying to slim this down and take out as much as possible to see if it's something else is causing the issue like group_vars, host_vars, var_files, etc...  I'll post shortly once I've slimmed down as much as possible with an update.

**Comment by digi691** [comment]
@jimi-c 

So I eliminated the var_files and handlers from the playbook.  It now looks like this:

``` yaml
---

- hosts: "{{ hosts|default('all') }}"
  strategy: "{{ strategy|default('linear') }}"
  user: root
  tasks:
    - name: test
      command: "echo foo"
```

This is the command I executed with Ansible 2.2.0 installed via pip
This is hitting a host list with 1573 hosts now

`$ ANSIBLE_DEBUG=1 ansible-playbook -i hosts -t sshkeys playbook-test.yml -f 100 -e strategy=linear`

I'll be sending you a DM in IRC with the link to the debug output from my dropbox account. When I did an strace there were a bunch of wait4()s  Let me know what else you need, etc...

**Comment by jimi-c** [comment]
Per further discussion on IRC, we're good to close this.

Thanks!
